File: Python/BinaryTreePreorderTraversal/BinaryTreePreorderTraversal.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    # (38ms) recursive
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def helper(root, nums):
            if not root:
                return
            nums.append(root.val)
            helper(root.left, nums)
            helper(root.right, nums)
        ans = []
        helper(root, ans)
        return ans

    # (42ms) iterative
    def preorderTraversal2(self, root):
        ans, stack = [], [root]
        while stack:
            root = stack.pop()
            if root:
                ans.append(root.val)
                # note that push right kid first
                stack.append(root.right)
                stack.append(root.left)
        return ans

File: JavaScript/BinaryTreePreorderTraversal/BinaryTreePreorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// (95ms) recursive
var preorderTraversal = function(root) {
    var helper = function(root, nums){
        if(root === null)
            return;
        nums.push(root.val);
        helper(root.left, nums);
        helper(root.right, nums);
    }
    var ans = [];
    helper(root, ans);
    return ans;
};

// (106ms) iterative
var preorderTraversal = function(root) {
    var stack = [];
    stack.push(root);
    var ans = [];
    while(stack.length > 0) {
        root = stack.pop();
        if(root !== null) {
            ans.push(root.val);
            stack.push(root.right);
            stack.push(root.left);
        }
    }
    return ans;
};